README.md
# Tanuki

A system for importing, storing, categorizing, browsing, displaying, and
searching assets, primarily images and videos. Attributes regarding the assets
are stored in a key-value store. Provides a simple web interface with basic
browsing and editing capabilities.

## Building and Testing

### Prerequisites

* [Rust](https://www.rust-lang.org) stable (2018 edition)
* [Flutter](https://flutter.dev) beta channel

### Building, Testing, Starting the Backend

```shell
$ cargo update
$ cargo build
$ cargo test
$ RUST_LOG=info cargo run --release
```

For more verbose debugging output, use `RUST_LOG=debug` in the command above.
For extremely verbose logging, use `RUST_LOG=trace` which will dump large
volumes of output.

### Building, Testing, Starting the Frontend

```shell
$ flutter pub get
$ flutter pub run environment_config:generate
$ flutter test
$ flutter run -d chrome
```

### environment_config

The frontend has some configuration that is set up at build time using the
[environment_config](https://pub.dev/packages/environment_config) package. The
generated file (`lib/environment_config.dart`) is not version controlled, and
the values can be set at build-time using either command-line arguments or
environment variables. See the `pubspec.yaml` for the names and the
`environment_config` README for instructions.

## Deploying

### Using Docker

The base directory contains a `docker-compose.yml` file which is used to build
the application in stages and produce a relatively small final image.

On the build host:

```shell
$ docker-compose build --pull --build-arg BASE_URL=http://192.168.1.1:3000
$ docker image rm 192.168.1.1:5000/tanuki_app
$ docker image tag tanuki_app 192.168.1.1:5000/tanuki_app
$ docker push 192.168.1.1:5000/tanuki_app
```

On the server, with a production version of the `docker-compose.yml` file:

```shell
$ docker-compose pull
$ docker-compose rm -f -s
$ docker-compose up -d
```

## Tools

### Finding Outdated Crates

Use https://github.com/kbknapp/cargo-outdated and run `cargo outdated`

### License checking

Use the https://github.com/Nemo157/cargo-lichking `cargo` utility. To install:

```shell
$ OPENSSL_ROOT_DIR=`brew --prefix openssl` \
  OPENSSL_LIB_DIR=`brew --prefix openssl`/lib \
  OPENSSL_INCLUDE_DIR=`brew --prefix openssl`/include \
  cargo install cargo-lichking
```

To get the list of licenses, and check for incompatibility:

```shell
$ cargo lichking list
$ cargo lichking check
```

However, need to look for "gpl" manually in the `list` output, as most licenses
that are compatible with MIT/Apache are also compatible with GPL.

## Architecture

Assets stored as-is in date/time formatted directory structure, metadata stored
in a key-value store, an HTTP server backend, and a single-page application for
a frontend.

* Backend written in [Rust](https://www.rust-lang.org)
* Front-end written in [Flutter](https://flutter.dev)
* Data store is [RocksDB](https://rocksdb.org)
* Wire protocol is [GraphQL](https://graphql.org) and HTTP

### Clean Architecture

The application is designed using the [Clean Architecture](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html) in which the application is divided into three layers: domain, data, and presentation. The domain layer defines the "policy" or business logic of the application, consisting of entities, use cases, and repositories. The data layer is the interface to the underlying system, defining the data models that are ultimately stored in a database. The presentation layer is what the user generally sees, the web interface, and to some extent, the GraphQL interface.

### Storage

When an asset is added to the system, several steps are performed:

1. A SHA256 checksum of the file contents is computed.
1. An identifier based on the import date/time, filename extension,
   and [ULID](https://github.com/ulid/spec) is generated.
1. A minimal document, with checksum and identifier, is stored in the database.
1. The asset is stored in a directory structure according to import date/time.
    - e.g. `2017/05/13/1630/01ce0d526z6cyzgm02ap0jv281.jpg`
    - the minutes are rounded to `:00`, `:15`, `:30`, or `:45`
    - the base filename is the ULID
    - the original file extension is retained

In previous versions of the application, assets were stored in a directory
structure reflecting the checksum of the asset, similar to the object store in
Git. There were two directory levels consisting of two pairs of leading digits
from the checksum, and the filename was the remainder of the checksum.

#### Benefits

* Assets are stored in directory structure reflecting time and order of addition
    - ULID sorts by time, so order of insertion is retained
* Number of directories and files at any particular level is reasonable
    - at most 96 directories per calendar day
    - files per directory limited to what can be processed within 15 minutes
* Can rebuild some of the metadata from the directory structure and file names
    - import date/time from file path
    - media type from extension
    - original date/time from file metadata
* Encoded path as asset ID allows serving asset without database lookup
    - base64 encoded asset path happens to be same length as SHA256
* Avoids filename collisions
    - names like `IMG_1234.JPG` easily collide with other devices

#### Drawbacks

The files are renamed, which might be a bother to some people. In many cases,
the file names are largely irrelevant, as most are of the form `IMG_1234.JPG`.
In other cases, the names are something ridiculous, like
`20150419171116-63EK7JXWKEVMDJVV-P1510081.jpg`, which encodes a date/time and
some seemingly random sequence of letters and numbers. The good news is the
original file name is recorded in the database.

## Project History

Original idea was inspired by [perkeep](https://perkeep.org) (née camlistore).
However, installing [Go](https://golang.org) on the server system in use at the
time, [Solaris](https://www.oracle.com/solaris/), was too difficult. Given the
operating system there would not be any readily available software to serve the
purpose. Would later learn that this application space is referred to as
"digital asset management."

### July 2014

[Python](https://www.python.org) script to ingest assets, insert records into
[CouchDB](http://couchdb.apache.org).
[Erlang](http://www.erlang.org)/[Nitrogen](http://nitrogenproject.com) interface
to display assets by one tag, in a long, single-column list.

### March 2015

Replaced the Python ingestion script with Erlang, no more Python in the main
application.

### January 2017

Replaced Erlang/Nitrogen code with [Elixir](https://elixir-lang.org) and
[Phoenix](https://phoenixframework.org); support querying by multiple tags;
pagination of assets; basic asset editor.

### March 2017

Replace static Elixir/Phoenix web pages with dynamic [Elm](http://elm-lang.org)
frontend, supporting multiple tags, locations, and years. Hides less frequently
used tags and locations by default, with expanders for showing everything. Form
input validation for asset edit page.

### November 2017

[Node.js](https://nodejs.org/) rewrite of the Elixir/Phoenix backend, using
[PouchDB](https://pouchdb.com) instead of CouchDB; Elm frontend still in place.

### March 2018

Replaced REST-like interface with [GraphQL](https://graphql.org).

Introduced automatic data migration to perform database schema upgrades.

### May 2018

Change asset storage design from sharding by SHA256 checksum to something akin
to Apple Photos (see above).

### October 2018

Replace frontend Elm code with [ReasonML](https://reasonml.github.io/en/).

### December 2019

Started to rewrite the Node.js backend in [Rust](https://www.rust-lang.org).

### February 2020

Started rewrite in [Dart](https://dart.dev) and [Flutter](https://flutter.dev)
with the intention of replacing all of the Node.js and ReasonML code. Would
later abandon using Dart (see architecture decision records).

### April 2020

Replaced backend Node.js code with [Rust](https://www.rust-lang.org).

### September 2020

Replaced frontend ReasonML code with [Flutter](https://flutter.dev).


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).
This file follows the convention described at
[Keep a Changelog](http://keepachangelog.com/en/1.0.0/).

## [1.0.0] - 2020-09-17
### Added
- Initial version


lib/assets.js
//
// Copyright (c) 2018 Nathan Fiedler
//
const config = require('config')
const dateFormat = require('dateformat')
const fs = require('fs-extra')
const path = require('path')
const logger = require('lib/logging')
const sharp = require('sharp')
const ffmpeg = require('fluent-ffmpeg')
const ffmpegStatic = require('ffmpeg-static')
ffmpeg.setFfmpegPath(ffmpegStatic.path)
const ffmpegProbe = require('ffprobe-static')
ffmpeg.setFfprobePath(ffmpegProbe.path)
const LRU = require('lru-cache')
const ULID = require('ulid')

//
// Code for operating on the stored assets.
//

const assetsPath = config.get('backend.assetPath')

const thumbnailCache = new LRU({
  max: 10485760,
  length: function (n, key) { return n.length }
})

/**
 * Convert the asset identifier to the file path of the asset.
 *
 * @param {string} assetId - asset identifier.
 * @returns {string} absolute path of the asset file.
 */
function assetPath (assetId) {
  const buf = Buffer.from(assetId, 'base64')
  const relpath = buf.toString('utf8')
  let fp = path.join(assetsPath, relpath)
  return path.isAbsolute(fp) ? fp : path.join(process.cwd(), fp)
}

/**
 * Convert the filename into a relative path where the asset will be located in
 * storage, and return as a base64 encoded value, suitable as an identifier.
 *
 * This is _not_ a pure function, since it involves both the current time as
 * will as a random number. It does, however, avoid any possibility of name
 * collisions.
 *
 * @param {Object} datetime - date/time of asset import, either a Date or a Number.
 * @param {string} filename - original name of the asset.
 * @returns {string} relative path of new asset, base64 encoded.
 */
function makeAssetId (datetime, filename) {
  // round the date/time down to the nearest quarter hour
  // (e.g. 21:50 becomes 21:45, 08:10 becomes 08:00)
  // this requires having a real Date object
  if (Number.isInteger(datetime)) {
    datetime = new Date(datetime)
  }
  const roundate = new Date(
    datetime.getFullYear(),
    datetime.getMonth(),
    datetime.getDate(),
    datetime.getHours(),
    Math.floor(datetime.getMinutes() / 15) * 15
  )
  const leadingPath = dateFormat(roundate, 'yyyy/mm/dd/HHMM')
  const name = ULID.ulid() + path.extname(filename)
  const relpath = path.join(leadingPath, name).toLowerCase()
  const buf = Buffer.from(relpath, 'utf8')
  return buf.toString('base64')
}

/**
 * Retrieve the thumbnail for the identified asset.
 * The image will be bounded to a box 240 by 240 pixels.
 *
 * @param {string} mimetype - the mimetype of the file.
 * @param {string} assetId - asset identifier.
 * @returns {Promise<Buffer>} promise resolving to an object holding
 *          the nodejs Buffer ('binary') and mime type ('mimetype');
 *          resolves to null if no thumbnail available.
 */
async function retrieveThumbnail (mimetype, assetId) {
  const cached = thumbnailCache.get(assetId)
  if (cached) {
    logger.info(`cache hit for ${assetId}`)
    return cached
  }
  logger.info(`cache miss for ${assetId}`)
  const filepath = assetPath(assetId)
  if (!fs.existsSync(filepath)) {
    logger.warn(`asset not found for ${assetId}`)
    return null
  }
  const thumbnail = await generateThumbnail(mimetype, filepath, 240, 240)
  if (thumbnail) {
    // only cache valid objects
    logger.debug(`cache store for ${assetId}`)
    thumbnailCache.set(assetId, thumbnail)
  }
  return thumbnail
}

/**
 * Generate a resized image for the given visual asset.
 *
 * @param {string} mimetype - the mimetype of the file.
 * @param {string} filepath - path of the file.
 * @param {number} width - width in pixels of the desired image.
 * @param {number} height - height in pixels of the desired image.
 * @returns {Promise<Buffer>} promise resolving to an object holding
 *          the nodejs Buffer ('binary') and mime type ('mimetype');
 *          resolves to null if no thumbnail available.
 */
async function generateThumbnail (mimetype, filepath, width, height) {
  if (width === null && height === null) {
    throw new Error('cannot have null width and height')
  }
  if (mimetype.startsWith('video/')) {
    return generateVideoThumbnail(filepath, width, height)
  } else if (mimetype.startsWith('image/')) {
    // fit the image into a box of the given size, convert to jpeg
    let bits = await sharp(filepath)
      .resize({
        width,
        height,
        fit: 'inside',
        withoutEnlargement: true
      })
      .toFormat('jpeg')
      .toBuffer()
    return {
      binary: bits,
      mimetype: 'image/jpeg'
    }
  } else {
    return null
  }
}

/**
 * Generate a thumbnail for the video asset.
 *
 * @param {string} filepath - path of the video asset.
 * @param {number} width - width in pixels of the desired image.
 * @param {number} height - height in pixels of the desired image.
 * @returns {Promise<Number>} promise resolving to an object holding
 *          the nodejs Buffer ('binary') and mime type ('mimetype');
 *          resolves to null if asset is missing.
 */
function generateVideoThumbnail (filepath, width, height) {
  // output file will remain on disk indefinitely, and be overwritten each
  // time, since the pixels value could be different; no sense keeping a
  // file for every possible pixels value either
  const dirname = path.dirname(filepath)
  const basename = path.basename(filepath, path.extname(filepath))
  const outfile = path.join(dirname, basename + '.jpg')
  let scale
  if (width && height) {
    scale = `w=${width}:h=${height}:force_original_aspect_ratio=decrease`
  } else if (width) {
    scale = `w=${width}:force_original_aspect_ratio=decrease`
  } else if (height) {
    scale = `h=${height}:force_original_aspect_ratio=decrease`
  }
  return new Promise((resolve, reject) => {
    ffmpeg(filepath)
      .on('end', function () {
        let binary = fs.readFileSync(outfile)
        resolve({
          binary,
          mimetype: 'image/jpeg'
        })
      })
      .on('error', function (err) {
        reject(err)
      })
      // There is the tempting screenshots() function, but unfortunately it does
      // not offer a means of maintaining the aspect ratio, and it insists on
      // writing the output as huge PNG files.
      //
      // see https://github.com/fluent-ffmpeg/node-fluent-ffmpeg/issues/776
      .outputOptions([
        '-vframes',
        '1',
        '-an',
        '-filter:v',
        `scale=${scale}`
      ])
      .save(outfile)
  })
}

/**
 * Generate a thumbnail image to fit a height of 300 pixels. This leaves the
 * width of the image to be determined by the original aspect ratio.
 *
 * @param {string} mimetype - the mimetype of the file.
 * @param {string} assetId - asset identifier.
 * @returns {Promise<Buffer>} promise resolving to an object holding
 *          the nodejs Buffer ('binary') and mime type ('mimetype');
 *          resolves to null if no thumbnail available.
 */
async function generateWideThumb (mimetype, assetId) {
  const filepath = assetPath(assetId)
  if (!fs.existsSync(filepath)) {
    logger.warn(`asset not found for ${assetId}`)
    return null
  }
  if (mimetype.startsWith('video/')) {
    return generateVideoThumbnail(filepath, null, 300)
  } else if (mimetype.startsWith('image/')) {
    // resize the image to the desired height, convert to jpeg
    let bits = await sharp(filepath)
      .resize(null, 300)
      .toFormat('jpeg')
      .toBuffer()
    return {
      binary: bits,
      mimetype: 'image/jpeg'
    }
  } else {
    return null
  }
}

/**
 * Generate a "preview" sized version of the identified asset.
 * The image will be bounded to a box 640 by 640 pixels.
 *
 * @param {string} mimetype - the mimetype of the file.
 * @param {string} assetId - asset identifier.
 * @returns {Promise<Buffer>} promise resolving to an object holding
 *          the nodejs Buffer ('binary') and mime type ('mimetype');
 *          resolves to null if asset is missing.
 */
async function generatePreview (mimetype, assetId) {
  let filepath = assetPath(assetId)
  if (!fs.existsSync(filepath)) {
    return null
  }
  return generateThumbnail(mimetype, filepath, 640, 640)
}

/**
 * Retrieve the duration of the asset, if it is a video, or null if not.
 *
 * @param {string} mimetype - the mimetype of the file.
 * @param {string} filepath - full path to the file.
 * @returns {Promise<Number>} promise resolving to the duration in seconds,
 *          or null if unable to extract data.
 */
async function getDuration (mimetype, filepath) {
  if (!fs.existsSync(filepath)) {
    return Promise.resolve(null)
  }
  if (mimetype.startsWith('video/')) {
    return new Promise((resolve, reject) => {
      ffmpeg.ffprobe(filepath, function (err, metadata) {
        if (err) {
          reject(err)
        } else {
          // streams: [{ codec_type: 'video', duration: '2.03'... }]
          try {
            for (let stream of metadata.streams) {
              if (stream.codec_type === 'video') {
                resolve(stream.duration)
              }
            }
            resolve(null)
          } catch (err) {
            resolve(null)
          }
        }
      })
    })
  }
  return Promise.resolve(null)
}

module.exports = {
  assetPath,
  generatePreview,
  generateWideThumb,
  getDuration,
  makeAssetId,
  retrieveThumbnail
}
